sw/device/tests/dif/dif_pwrmgr_smoketest.c

```diff
@@ -0,0 +1,80 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/base/mmio.h"
+#include "sw/device/lib/dif/dif_pwrmgr.h"
+#include "sw/device/lib/runtime/log.h"
+#include "sw/device/lib/testing/check.h"
+#include "sw/device/lib/testing/test_main.h"
+#include "sw/device/lib/usbdev.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"  // Generated.
+
+static dif_pwrmgr_t pwrmgr;
+
+const test_config_t kTestConfig;
+const dif_pwrmgr_wakeup_reason_t test_wakeup_reason = {
+    .types = kDifPwrmgrWakeupTypeRequest,
+    .request_sources = kDifPwrmgrWakeupRequestSourceThree,
+};
+
+const dif_pwrmgr_wakeup_reason_t por_wakeup_reason = {
+    .types = 0,
+    .request_sources = 0,
+};
+
+static bool compare_wakeup_reasons(const dif_pwrmgr_wakeup_reason_t *lhs,
+                                   const dif_pwrmgr_wakeup_reason_t *rhs) {
+  return lhs->types == rhs->types &&
+         lhs->request_sources == rhs->request_sources;
+}
+
+bool test_main(void) {
+  bool test_passed = false;
+
+  CHECK(dif_pwrmgr_init(
+            (dif_pwrmgr_params_t){
+                .base_addr =
+                    mmio_region_from_addr(TOP_EARLGREY_PWRMGR_AON_BASE_ADDR),
+            },
+            &pwrmgr) == kDifPwrmgrOk);
+
+  // Assuming the chip hasn't slept yet, wakeup reason should be empty.
+  dif_pwrmgr_wakeup_reason_t wakeup_reason;
+  CHECK(dif_pwrmgr_wakeup_reason_get(&pwrmgr, &wakeup_reason) == kDifPwrmgrOk);
+
+  if (compare_wakeup_reasons(&wakeup_reason, &por_wakeup_reason)) {
+    LOG_INFO("Powered up for the first time, begin test");
+  } else if (compare_wakeup_reasons(&wakeup_reason, &test_wakeup_reason)) {
+    test_passed = true;
+    LOG_INFO("USB wakeup detected");
+  } else {
+    CHECK(false);
+  }
+
+  // Fake low power entry through usb
+  // Force usb to output suspend indication
+  if (!test_passed) {
+    usbdev_wake(true);
+    usbdev_force_suspend();
```

Comment:
Do you have a fake PHY that responds to this?

The signal will affect the AON wake logic but I'm not quite sure what keeps the input from D+,D- in the suspend state?

(This signal is intended to tell the PHY it can go into a low power state because the link is suspended so just single ended D+/D- changes at low speed are important)

Reply:
this is really hoaky and only works in DV (i imagine).
Basically the DV enviornment (perhaps incorrectly) is always driving 1/0 on USBP/N.  So basically as long as I have "something" different, it'll cause a wakeup.  That's basically all that I'm relying on.

Reply:
Ok, that is fine as a quick DV check. Might be worth a comment about what you are expecting.